common/src/main/java/com/wynntils/core/config/ConfigManager.java

```diff
@@ -37,61 +37,51 @@ public final class ConfigManager extends Manager {
 
-    private static final List<ConfigHolder> OVERLAY_GROUP_CONFIG_HOLDERS = new ArrayList<>();
-    private static final List<OverlayGroupHolder> OVERLAY_GROUP_FIELDS = new ArrayList<>();
-
     private final File userConfig;
     private JsonObject configObject;
 
-    public ConfigManager(ConfigUpfixerManager configUpfixerManager, JsonManager jsonManager) {
-        super(List.of(configUpfixerManager, jsonManager));
+    public ConfigManager(
+            ConfigUpfixerManager configUpfixerManager,
+            JsonManager jsonManager,
+            FeatureManager feature,
+            OverlayManager overlay) {
+        super(List.of(configUpfixerManager, jsonManager, feature, overlay));
 
         userConfig = new File(CONFIGS, McUtils.mc().getUser().getUuid() + FILE_SUFFIX);
+    }
 
+    public void init() {
         // First, we load the config file
         configObject = Managers.Json.loadPreciousJson(userConfig);
 
         // Now, we have to apply upfixers, before any config loading happens
-        if (configUpfixerManager.runUpfixers(configObject)) {
+        if (Managers.ConfigUpfixer.runUpfixers(configObject)) {
             Managers.Json.savePreciousJson(userConfig, configObject);
         }
-    }
-
-    public void registerFeature(Feature feature) {
-        registerOverlayGroups(feature);
-
-        for (Overlay overlay : feature.getOverlays()) {
-            registerConfigOptions(overlay);
-        }
 
-        registerConfigOptions(feature);
-    }
+        // Register all features and overlays
+        Managers.Feature.getFeatures().forEach(this::registerFeature);
 
-    private void registerOverlayGroups(Feature feature) {
-        List<OverlayGroupHolder> holders = Stream.of(feature.getClass().getDeclaredFields())
-                .filter(f -> f.isAnnotationPresent(OverlayGroup.class))
-                .map(field -> new OverlayGroupHolder(
-                        field, feature, field.getAnnotation(OverlayGroup.class).instances()))
-                .toList();
+        // Finish off the config init process
 
-        if (holders.isEmpty()) return;
+        // Load configs for all features
+        Managers.Config.reloadConfiguration();
 
-        OVERLAY_GROUP_FIELDS.addAll(holders);
-        feature.addOverlayGroups(holders);
+        // Save config file after loading all configurables' options
+        Managers.Config.saveConfig();
 
-        for (OverlayGroupHolder holder : holders) {
-            holder.initGroup(
-                    IntStream.rangeClosed(1, holder.getDefaultCount()).boxed().toList());
-
-            List<ConfigHolder> overlayHolders = holder.getOverlays().stream()
-                    .map(this::getConfigOptions)
-                    .flatMap(List::stream)
-                    .toList();
-
-            holder.getOverlays().forEach(overlay -> overlay.addConfigOptions(this.getConfigOptions(overlay)));
+        // Create default config file containing all configurables' options
```

Comment:
I'm honestly not sure what the default config file does anymore. I think I/we had some kind of idea about this, but it kind of never panned out, right?

And if we do want to keep it, don't you have to save the default file from the Configs as they are before you load in the users config file?

Reply:
The default config file is a config file, containing all keys, with their default values, as examples. And no, the config holders know the default value, because it is needed for reseting configs. So this is fine here.

Reply:
Ah, when you create the config holder from the annotation you read the default value? Clever.